src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java

```diff
@@ -865,8 +866,18 @@ public void blockedASTTransformsGrab() throws Exception {
     }
 
     private static Object evaluate(Whitelist whitelist, String script) {
-        GroovyShell shell = new GroovyShell(GroovySandbox.createSecureCompilerConfiguration());
-        Object actual = GroovySandbox.run(shell, script, whitelist);
+        CompilerConfiguration cc = GroovySandbox.createSecureCompilerConfiguration();
+        GroovyShell shell = new GroovyShell(cc);
+        try {
+            Field loaderF = GroovyShell.class.getDeclaredField("loader");
```

Comment:
Is there a way some of the subtle code here can be shared with `SecureGroovyScript.evaluate`? Is that what you were proposing in https://github.com/jenkinsci/script-security-plugin/pull/259#discussion_r294446731?

Reply:
Yeah, I think we could create something like `static void setGroovyClassLoader(GroovyShell, ClassLoader)` that sets the loader field after wrapping the passed loader in CleanGroovyClassLoader, and then update both places to call that method. 

Maybe possible to go further and introduce a new subtype of `GroovyShell` that does that in the constructor so that we could also remove the duplication in [`CpsGroovyShell`](https://github.com/jenkinsci/workflow-cps-plugin/blob/7eab5dc0c39289a372a05904d65596c4192c4005/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsGroovyShell.java#L50-L57) (might need to tweak the memory leak fixes in workflow-cps if we did do that). We could also add a `GroovyShell prepareSecureGroovyShell(ClassLoader, Binding, CompilerConfiguration)` in `GroovySandbox` and recommend downstream users of the plugin use that instead of instantiating `GroovyShell` directly.

Reply:
Could all be deferred to some follow-up. Maybe capture the conversation in a TODO comment or something.

Reply:
TODO added in 9e8a1ae.